### PR TITLE
🏗✨✅ Add tests for `gulp serve --new_server` to CI workflow

### LINKED ISSUE #27339: Add esm (module / no-module) support to `gulp serve`
The current `gulp serve` server does not support esm (module / no-module). Make it so that it's possible to test esm builds during development.

### PR DESCRIPTION
**Pull request by rsimha**
In #27340, `gulp serve` was updated with a set of typescript transforms and a `--new_server` flag.

This PR addresses https://github.com/ampproject/amphtml/pull/27340#issuecomment-602860266, and does the following:
- Migrates test files for server transforms from #27219
- Adds a `README.md` with instructions for how to add tests
- Adds a new task `gulp server-tests` to run server tests
- Adds a new `SERVER` build target that runs these tests when server code changes

**To run these tests:**
```
gulp server-tests
```
**Passing tests:**
image

**Failing tests:**
image

Follow up to #27340
Fixes #27339

Co-authored by @kristoferbaxter

### CODE REVIEW

File: build-system/server/new-server/transforms/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+## Custom transforms for `gulp serve`.
+
+**Parent directory:** `build-system/server/new-server/transforms/`
+
+**Structure:** Each subdirectory should contain the following:
+
+- A single `.ts` file that implements a specific transform (`*-transform.ts`)
+- One or more pairs of input and output files for the transform (`test-*-[input|output].html`)
+
+**Example:**
```
Comment (kristoferbaxter) [inline build-system/server/new-server/transforms/README.md]: This is a great description and addition. Thank you for doing this!

File: build-system/tasks/server-tests.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  console.groupEnd();
+}
+
+/**
+ * Runs tests for a single input file
+ *
+ * @return {!ReadableStream}
+ */
+function runTest() {
+  return through.obj(async (file, enc, cb) => {
```
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/server-tests.js]: Our compilation toolchain heavily relies on `gulp.src` and multiple other `gulp` plugins (e.g. closure compiler). In those cases, I believe the use of `gulp` actually reduces the complexity of our code.

I do acknowledge though, that this task could've been written as a simple node script without the use of `gulp`. I chose to do so in order to follow established convention in our toolchain, and to benefit from showing up in places like `gulp help`.

Re: migrating the entire toolchain away from `gulp`, it's going to be a significantly long project given the sheer surface area (and definitely not a single PR's worth of work). Therefore, I'm not confident that we will ever be able to prioritize a full rewrite.

File: build-system/server/new-server/transforms/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+```
+└── transforms
+    ├── foo
+    │   ├── foo-transform.ts
+    │   ├── test-name-input.html
+    │   └── test-name-output.html
+    ├── bar
+    │   ├── bar-transform.ts
+    │   ├── test-name1-input.html
```
[the remark this answers is not in the document]
Comment (erwinmombay) [reply build-system/server/new-server/transforms/README.md]: +1

[the remark this answers is not in the document]
Comment (erwinmombay) [reply build-system/server/new-server/transforms/README.md]: maybe we could follow the jest test harness structure? not sure if thats *too* much, but the pattern is already familiar to us

File: build-system/tasks/server-tests.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    const testName = path
+      .relative(transformsDir, inputFile)
+      .replace('-input.html', '');
+    const outputFile = inputFile.replace('input.html', 'output.html');
+    const dirName = path.dirname(inputFile);
+    const tsTransform = globby.sync(path.resolve(dirName, '*.ts'))[0];
+    const jsTransform = tsTransform
+      .replace('/transforms/', '/transforms/dist/')
+      .replace('.ts', '.js');
+    const input = fs.readFileSync(inputFile, 'utf8');
```
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/server-tests.js]: Done.

File: build-system/tasks/server-tests.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    const expected = fs.readFileSync(outputFile, 'utf8');
+    const transform = require(jsTransform).default;
```
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/server-tests.js]: When I tried using `import`, it returned this error:
```
(node:94008) UnhandledPromiseRejectionWarning: Error: Not supported
```
I believe this [will work](https://stackoverflow.com/questions/37132031/nodejs-plans-to-support-import-export-es6-es2015-modules) once node LTS moves from 12 to 14, and I agree we should change this to `import` at that point.

File: build-system/tasks/server-tests.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+ */
+function runTest() {
+  return through.obj(async (file, enc, cb) => {
+    const inputFile = file.path;
+    const testName = path
+      .relative(transformsDir, inputFile)
+      .replace('-input.html', '');
+    const outputFile = inputFile.replace('input.html', 'output.html');
+    const dirName = path.dirname(inputFile);
+    const tsTransform = globby.sync(path.resolve(dirName, '*.ts'))[0];
```
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/server-tests.js]: Good idea, done.

File: build-system/server/new-server/transforms/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/server/new-server/transforms/README.md]: I initially flattened the directory structure for simplicity, but I like your idea better. I've reinstated the `test/` subdirectory.

File: build-system/tasks/server-tests.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+ *
+ * @return {!ReadableStream}
+ */
+function runTest() {
+  return through.obj(async (file, enc, cb) => {
+    const inputFile = file.path;
+    const testName = path
+      .relative(transformsDir, inputFile)
+      .replace('-input.html', '');
+    const outputFile = inputFile.replace('input.html', 'output.html');
```
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/server-tests.js]: Totally agree. I've refactored all the path functionality into standalone helper functions.

**Pull request merged**